src/audio/pipeline/pipeline-stream.c

```diff
@@ -268,24 +270,123 @@ static void pipeline_trigger_xrun(struct pipeline *p, struct comp_dev **host)
 	} while (true);
 }
 
+#if CONFIG_KCPS_DYNAMIC_CLOCK_CONTROL
+static int add_pipeline_cps_consumption(struct comp_dev *current,
+					struct comp_buffer *calling_buf,
+					struct pipeline_walk_context *ctx, int dir)
+{
+	struct pipeline_data *ppl_data = ctx->comp_data;
+	struct ipc4_base_module_cfg *cd  = NULL;
+
+	pipe_dbg(ppl_data->p, "add_pipeline_cps_consumption(), current->comp.id = %u, dir = %u",
+		 dev_comp_id(current), dir);
+
+	if (!comp_is_single_pipeline(current, ppl_data->start)) {
+		pipe_dbg(ppl_data->p, "add_pipeline_cps_consumption(), current is from another pipeline");
+		return 0;
+	}
+
+	/* complete component init */
+	current->pipeline = ppl_data->p;
+
+	/* modules created throug module adapter have different priv_data */
+	if (current->drv->type != SOF_COMP_MODULE_ADAPTER) {
+		cd = comp_get_drvdata(current);
+	} else {
+		struct processing_module *mod = comp_get_drvdata(current);
+		struct module_data *md = &mod->priv;
+		cd = &md->cfg.base_cfg;
+	}
+
+	if (cd->cpc == 0) {
+		/* Use maximum clock budget, assume 1ms chunk size */
+		cd->cpc = CLK_MAX_CPU_HZ / 1000;
+		tr_warn(pipe,
+			"0 CPS requested for module: %#x, core: %d using safe max CPC: %u",
+			current->ipc_config.id, ppl_data->p->core, cd->cpc);
+	}
+
+	int kcps = cd->cpc * 1000 / ppl_data->p->period;
+	tr_info(pipe, "Registering KCPS consumption: %d, core: %d", kcps, ppl_data->p->core);
+	core_kcps_adjust(ppl_data->p->core, kcps);
+
+	int summary_cps = core_kcps_get(ppl_data->p->core);
+	tr_info(pipe, "Sum of KCPS consumption: %d, core: %d", summary_cps, ppl_data->p->core);
+
+	return pipeline_for_each_comp(current, ctx, dir);
+}
+
+static int remove_pipeline_cps_consumption(struct comp_dev *current,
+					   struct comp_buffer *calling_buf,
+					   struct pipeline_walk_context *ctx, int dir)
+{
+	struct pipeline_data *ppl_data = ctx->comp_data;
+	struct ipc4_base_module_cfg *cd  = NULL;
+
+	pipe_dbg(ppl_data->p, "remove_pipeline_cps_consumption(), current->comp.id = %u, dir = %u",
+		 dev_comp_id(current), dir);
+
+	if (!comp_is_single_pipeline(current, ppl_data->start)) {
+		pipe_dbg(ppl_data->p, "remove_pipeline_cps_consumption(), current is from another pipeline");
+		return 0;
+	}
+
+	/* complete component init */
+	current->pipeline = ppl_data->p;
+
+	/* modules created throug module adapter have different priv_data */
+	if (current->drv->type != SOF_COMP_MODULE_ADAPTER) {
+		cd = comp_get_drvdata(current);
+	} else {
+		struct processing_module *mod = comp_get_drvdata(current);
+		struct module_data *md = &mod->priv;
+		cd = &md->cfg.base_cfg;
+	}
+
+	int kcps = cd->cpc * 1000/ppl_data->p->period;
+	core_kcps_adjust(ppl_data->p->core, -kcps); /* 1000 chunks per second, so cpc value matches kcps? */
+	tr_info(pipe, "Unregistering KCPS consumption: %d, core: %d", kcps, ppl_data->p->core);
+	int summary_cps = core_kcps_get(ppl_data->p->core);
+	tr_info(pipe, "Sum of KCPS consumption: %d, core: %d", summary_cps, ppl_data->p->core);
```

Comment:
can this print be moved within `core_kcps_adjust()` with some adjustment?

Reply:
EDIT: reverted and moved back to pipeline stream because of never ending compile errors